File: lang/ruby/src/fun/vac/algorithms/sorting/MergeSort.rb
module MergeSort

    def self.merge(arr, lo, mid, hi)
        if lo == mid
            return
        end

        merge(arr, lo, (lo + mid) >> 1, mid)
        merge(arr, mid, (mid + hi) >> 1, hi)

        m = lo
        n = mid

        sorted = Array.new(hi - lo)

        (0 ... sorted.length).each do |i|
            if m != mid && (n == hi || arr[m] < arr[n])
                sorted[i] = arr[m]
                m += 1
            else
                sorted[i] = arr[n]
                n += 1
            end
        end

        cursor = 0
        (lo ... hi).each do |i|
            arr[i] = sorted[cursor]
            cursor += 1
        end
    end

    def self.sort(arr)
        merge(arr, 0, arr.length >> 1, arr.length)
    end
end

File: lang/php/src/vac.fun/algorithms/sorting/MergeSort.php
<?php
    namespace vac\fun\algorithms\sorting\MergeSort;

    function merge(array &$arr, int $lo, int $mid, int $hi): void {
        if ($lo == $mid) {
            return;
        }

        merge($arr, $lo, ($lo + $mid) >> 1, $mid);
        merge($arr, $mid, ($mid + $hi) >> 1, $hi);

        $m = $lo;
        $n = $mid;

        $sorted = array_pad(array(), $hi - $lo, 0);

        foreach ($sorted as &$v) {
            if ($m != $mid && ($n == $hi || $arr[$m] < $arr[$n])) {
                $v = $arr[$m];
                $m++;
            } else {
                $v = $arr[$n];
                $n++;
            }
        }

        $cursor = 0;
        for ($i = $lo; $i < $hi; $i++) {
            $arr[$i] = $sorted[$cursor];
            $cursor++;
        }
    }

    function sort(array &$arr): void {
        merge($arr, 0, count($arr) >> 1, count($arr));
    }
?>
